crypto/src/proofs/transparent.rs

```diff
@@ -587,6 +565,217 @@ impl TryFrom<transparent_proofs::SwapProof> for SwapProof {
     }
 }
 
+/// Transparent proof for swap creation.
+///
+/// Swaps create an output NFT encoding the swap data so they are most similar to Output operations.
+///
+/// This structure keeps track of the auxiliary (private) inputs.
+#[derive(Clone, Debug)]
+pub struct SwapProof {
+    // The diversified base for the destination address.
+    pub g_d: decaf377::Element,
+    // The transmission key for the destination address.
+    pub pk_d: ka::Public,
+    // The value of asset 1 in the swap.
+    pub value_t1: Value,
+    // The value of asset 2 in the swap.
+    pub value_t2: Value,
+    // The fee value associated with the swap.
+    // TODO: should always be of penumbra token type, maybe this should
+    // be uint64 or a hardcoded value?
+    pub value_fee: Value,
+    // The asset ID of the Swap NFT.
+    pub swap_nft_asset_id: asset::Id,
+    // The blinding factor used for generating the value commitment.
+    // TODO: can this be shared amongst t1/t2/fee or is this unsafe?
+    pub v_blinding: Fr,
+    // The blinding factor used for generating the note commitment for the Swap NFT.
+    pub note_blinding: Fq,
+    // The ephemeral secret key that corresponds to the public key.
+    pub esk: ka::Secret,
+}
+
+impl SwapProof {
+    /// Called to verify the proof using the provided public inputs.
+    ///
+    /// The public inputs are:
+    /// * value commitment of the asset 1's contribution to the transaction,
+    /// * value commitment of the asset 2's contribution to the transaction,
+    /// * value commitment of the fee's contribution to the transaction,
+    /// * note commitment of the new swap NFT note,
+    /// * the ephemeral public key used to generate the new swap NFT note.
+    pub fn verify(
+        &self,
+        value_1_commitment: value::Commitment,
+        value_2_commitment: value::Commitment,
+        value_fee_commitment: value::Commitment,
+        note_commitment: note::Commitment,
+        epk: ka::Public,
+    ) -> anyhow::Result<(), Error> {
+        // Note commitment integrity.
+        let s_component_transmission_key = Fq::from_bytes(self.pk_d.0);
+        if let Ok(transmission_key_s) = s_component_transmission_key {
+            // Checks the note commitment of the Swap NFT.
+            let note_commitment_test = note::commitment(
+                self.note_blinding,
+                Value {
+                    // The swap NFT is always amount 1.
+                    amount: 1,
+                    asset_id: self.swap_nft_asset_id,
+                },
+                self.g_d,
+                transmission_key_s,
+            );
+
+            if note_commitment != note_commitment_test {
+                return Err(Error::NoteCommitmentMismatch);
```

Comment:
note: we could just as well use `anyhow` here, if it's easier -- since the transparent proofs are supposed to model ZK proofs, we don't need to do fine-grained matching on the error cases, since all we'll get is an opaque "failed to verify" error anyways.